### PR TITLE
extension attributes + photo

### PR DESCRIPTION
**Pull request by frantiekm**
We would like to contribute 2 changes:
- extension attributes management on user object 
- photo attribute management for user object

### CODE REVIEW

**Comment by matusmacik** [comment]
Hello, thank you for your pull request. Would it be possible to rebase the request with the current state of the project and re-test the project with the integration tests that are present? Also please add a more detailed description of the contributions if possible.  

Thank you.

**Comment by frantiekm** [comment]
I merge the changes I will do some testing and let you know the results. As for the details we simply added logic to able to manage user photo in MS Graph API. And we also aded following attributes:
- employeeHireDate
- employeeLeaveDateTime
- employeeType
- faxNumber
- employeeId
- extensionAttribute1-15
- photo

**Comment by frantiekm** [comment]
So I tested the merged connector and it seems fine + I run all testes with success

File: .github/workflows/release.yml
Comment (matusmacik) [inline .github/workflows/release.yml:1]: Please do not add .github workflow actions, maven run configuration (or any such configuration files) in the contribution pull requests.

File: .run/connector-msgraph [clean,install...].run.xml
Comment (matusmacik) [inline .run/connector-msgraph [clean,install...].run.xml:1]: Please do not add .github workflow actions, maven run configuration (or any such configuration files) in the contribution pull requests.

File: CHANGELOG.md
Comment (matusmacik) [inline CHANGELOG.md:1]: Please do not add changelog files, an appropriate commit message should be enough.

File: src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java [hunk 1]
```diff
@@ -376,6 +387,9 @@ protected ObjectClassInfo objectClassInfo() {
         ;
         userObjClassBuilder.addAttributeInfo(attrInterests.build());
 
+        AttributeInfoBuilder userPhoto = new AttributeInfoBuilder(ATTR_USERPHOTO);
```
Comment (matusmacik) [inline src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java:390]: Please add the setReturnedByDefault(false) property for such an attribute, the attribute might not be needed in most deployments of the connector and additionally, it requires an additional roundtrip to the remote service.

File: src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java [hunk 2]
```diff
@@ -387,6 +401,14 @@ protected ObjectClassInfo objectClassInfo() {
         attrMail.setRequired(false).setType(String.class).setCreateable(true).setUpdateable(true).setReadable(true);
         userObjClassBuilder.addAttributeInfo(attrMail.build());
 
+        // UPDATE Extension ATTRIBUTES
+        for (int i = 1; i <= NUMBER_OF_EXTENSIONS; i++) {
```
Comment (matusmacik) [inline src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java:405]: Please add the setReturnedByDefault(false) property for such an attribute, the attribute might not be needed in most deployments of the connector and additionally, it requires an additional roundtrip to the remote service. Also, please add the conditionals needed regarding the retrieval of the property which check "ATTRS_TO_GET" operation options in the correct places later in code.

File: src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java (hunk 1 above)
Comment (matusmacik) [reply src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java:390]: Also, please add the conditionals needed regarding the retrieval of the property which check "ATTRS_TO_GET" operation options in the correct places later in code.

File: src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java [hunk 3]
```diff
@@ -1087,6 +1162,10 @@ protected boolean handleJSONObject(OperationOptions options, JSONObject user, Re
             user = saturateRoleMembership(user);
         }
 
+        if (!Boolean.TRUE.equals(options.getAllowPartialAttributeValues())) {
```
Comment (matusmacik) [inline src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java:1165]: this seems an appropriate place to also add a check for the "ATTRS_TO_GET" operation options regarding the 'photo' parameter so there is no additional roundtrip.

File: src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java [hunk 4]
```diff
@@ -1270,32 +1373,97 @@ public String getSelectorSingle(OperationOptions options) {
         if (options != null) {
 
             return selector(getSchemaTranslator().filter(ObjectClass.ACCOUNT_NAME, options,
```
Comment (matusmacik) [inline src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java:1375]: please avoid linebreak and formatting changes

File: src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java [hunk 5]
```diff
@@ -1270,32 +1373,97 @@ public String getSelectorSingle(OperationOptions options) {
...
-                    ATTR_GIVENNAME, ATTR_IMADDRESSES, ATTR_ID,
-                    ATTR_JOBTITLE, ATTR_MAIL, ATTR_MOBILEPHONE, ATTR_OFFICELOCATION,
-                    ATTR_ONPREMISESLASTSYNCDATETIME, ATTR_ONPREMISESSECURITYIDENTIFIER,
-                    ATTR_ONPREMISESSYNCENABLED, ATTR_PASSWORDPOLICIES,
-                    ATTR_POSTALCODE, ATTR_PREFERREDLANGUAGE,
-                    ATTR_PROXYADDRESSES, ATTR_STATE, ATTR_STREETADDRESS, ATTR_SURNAME,
-                    ATTR_USAGELOCATION, ATTR_USERTYPE, ATTR_ASSIGNEDLICENSES,
-                    ATTR_EXTERNALUSERSTATE, ATTR_EXTERNALUSERSTATECHANGEDATETIME, ATTR_MANAGER);
-
+            return selector(
```
Comment (matusmacik) [inline src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java:1427]: please avoid linebreak and formatting changes

File: src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java [hunk 6]
```diff
@@ -1270,32 +1373,97 @@ public String getSelectorSingle(OperationOptions options) {
...
-                    ATTR_GIVENNAME, ATTR_HIREDATE, ATTR_IMADDRESSES, ATTR_ID, ATTR_INTERESTS,
-                    ATTR_JOBTITLE, ATTR_MAIL, ATTR_MOBILEPHONE, ATTR_MYSITE, ATTR_OFFICELOCATION,
-                    ATTR_ONPREMISESLASTSYNCDATETIME, ATTR_ONPREMISESSECURITYIDENTIFIER,
-                    ATTR_ONPREMISESSYNCENABLED, ATTR_PASSWORDPOLICIES, ATTR_PASTPROJECTS,
-                    ATTR_POSTALCODE, ATTR_PREFERREDLANGUAGE, ATTR_PREFERREDNAME,
-                    ATTR_PROXYADDRESSES, ATTR_RESPONSIBILITIES, ATTR_SCHOOLS,
-                    ATTR_SKILLS, ATTR_STATE, ATTR_STREETADDRESS, ATTR_SURNAME,
-                    ATTR_USAGELOCATION, ATTR_USERTYPE, ATTR_ASSIGNEDLICENSES,
-                    ATTR_EXTERNALUSERSTATE, ATTR_EXTERNALUSERSTATECHANGEDATETIME, ATTR_MANAGER));
+                    ATTR_ACCOUNTENABLED,
```
Comment (matusmacik) [inline src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java:1376]: please avoid linebreak and formatting changes

**Review by matusmacik — COMMENTED** [review]
Hello, I made an intermediary review of the code contribution, please have a look at the comment sections, consider the appropriate changes, and issue the modified pull request.
 Thank you.

File: .github/workflows/release.yml
Comment (frantiekm) [reply .github/workflows/release.yml:1]: done

File: .run/connector-msgraph [clean,install...].run.xml
Comment (frantiekm) [reply .run/connector-msgraph [clean,install...].run.xml:1]: done

File: pom.xml
[the remark this answers is not in the document]
Comment (frantiekm) [reply pom.xml:1]: done, chanegd back to 1.1.3.0

File: CHANGELOG.md
Comment (frantiekm) [reply CHANGELOG.md:1]: deleted

File: src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java (hunk 1 above)
Comment (frantiekm) [reply src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java:390]: both done

File: src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java (hunk 2 above)
Comment (frantiekm) [reply src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java:405]: setReturnedByDefault(false), ATTRS_TO_GET in progress

File: src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java (hunk 3 above)
Comment (frantiekm) [reply src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java]: done

File: src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java (hunk 5 above)
Comment (frantiekm) [reply src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java:1393]: done

File: src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java (hunk 6 above)
Comment (frantiekm) [reply src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java]: done

File: src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java (hunk 2 above)
Comment (frantiekm) [reply src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java:405]: extension attributes are return by default (if you add them to selector) and they don't require additional round trip. So I don't see necessity of checking them against ATTR_TO_GET as the result would be another selector. So I personaly find this irrelevant. Can we agree on this? @matusmacik

Comment (matusmacik) [reply src/main/java/com/evolveum/polygon/connector/msgraphapi/UserProcessing.java:405]: Hello, thank you. Yes in that case we can leave them in the schema.

**Comment by matusmacik** [comment]
Thank you for updating the code and overall for the contribution, I will proceed with the merge.

**Pull request merged**